keps/sig-network/1111-dynamically-re-configure-lb-node-set/README.md

```diff
@@ -0,0 +1,681 @@
+load balancer drop incoming traffic. This change can take up to minutes to get
+applied completely and thus has important impacts on the service SLAs linked to
+the ingress of the service.
+
+## Motivation
+
+The motivation of this KEP is to design a mechanism that allows for the load
+balancer set to change without re-configuring the load balancer when any
+__existing__ nodes transition in/out of the set. Emphasis is being placed on
+"existing", since nodes which are being fully deleted or newly added to the
+cluster will still require a re-configuration (following what was detailed in
+[summary](#summary)) of the load balancer by the KCCM.
+
+To be able to fully address this problem, we will also need to discuss a
+question which remains un-answered, namely: what should the "inclusion
+predicates" be, for any node to get added to / removed from the set?
+
+Addressing and answering the problem presented above will allow the Kubernetes
+cluster to react much more efficiently to transient errors, for what application
+load balancers are concerned. This will increase application ingress SLAs and
+de-couple it from the "level of entropy" impacting circumstantial components /
+constructs which might be orthogonal to the performance of the application
+itself.  
+
+### Goals
+
+#### Defining "healthy"
+
+- Answer the question: what constitutes a "healthy" node from the point of view
+  of an __application load balancer__? Under which conditions should a node be /
+  not be considered as an acceptable next hop for load balancer traffic? Should
+  a healthy __node__ even be the guiding criteria determining where an
+  __application__ load balancer should send traffic?
+
+- Improve the performance of keeping the cloud provider load balancers in sync
+  with the state of the Kubernetes cluster, by dynamically having the load
+  balancer identify the set of nodes to be used, solely through the utilization
+  of their health check (HC) probes.  
+
+#### Align health checks between services
+
+- Reduce service downtime caused by updating fields on the service object which
+  do not have a direct correlation to the load balancer but more so to its HC /
+  where traffic should be forwarded, ex: updating the `externalTrafficPolicy`
+  field from `Cluster` to `Local` or vice versa. Such changes to the service
+  object should not end up re-configuring the load balancer, only change the
+  response returned by the service proxy to the load balancer's HC.
+
+```
+<<[UNRESOLVED HC-linked-service-fields ]>>
+what are all the fields on the service object which fulfill this statement?
+<<[/UNRESOLVED]>>
+```
+
+### Non-Goals
+
+- Avoiding re-configuration of the load balancer for fully deleted / newly added
+  cluster nodes.
+
+## Proposal
+
+Up until this point the KCCM has defined the following conditions for what is
+considered a "healthy" node:
+
+1. Is the node not tainted with the key: `ToBeDeletedByClusterAutoscaler`?
```

Comment:
Looking at https://github.com/kubernetes/autoscaler/blob/8dcd3b3eb2d1d28b7ba05a3af48a171cb40547b9/cluster-autoscaler/core/scaledown/actuation/actuator.go#L72-L106 it seems that the assertion being made is that this taint goes on nodes which are ALREADY EMPTY, so we can/should treat this the same as a node deletion, I think?

Reply:
I am actually going to revise my statements with regards to this taint, yet again....

I am not sure how required it is. But I believe the reasons for adding it is this: 

We know that any node transitioning in / out of the LB set gets added / removed by means of having the CCM call:

```go
EnsureLoadBalancer(ctx context.Context, clusterName string, service *v1.Service, nodes []*v1.Node) (*v1.LoadBalancerStatus, error)
```   

Now, if one node gets tainted and subsequently removed from the set this isn't a problem - because we re-configure the set with the remaining nodes, so I don't really see how that could create any problem. However, if the auto-scaler scales down multiple nodes, I can definitely see the following happening: 

1. node1, node2 are to be scaled down and remove 
2. node1, node2 are tainted
3. node1, node2 are drained, etc...
4. node1 associated instance is deleted
5. node1 `Node` is deleted 
6. CCM re-configures LB set with all nodes + node2 
7. node2 instance is deleted 
8. CCM starts erroring because node2 instance cannot be found 
9. node2 `Node` is deleted
10. CCM re-configures LB set with all nodes
11. LB is now correctly configured. 

This really matches the description painted on the original PR, https://github.com/kubernetes/kubernetes/pull/105946#issue-1037895311

```
If multiple such nodes are being deleted, the list of nodes passed to cloud-provider code via UpdateLoadBalancer() will be stale, causing a few failed updates before the backends stabilize.
```

I am not sure how much we care about the CCM reconciling some incorrect _transient_ state however, that is kind-of the "status quo of cloud"  - things might not be right, right now, but it will eventually get there. However, maybe the better thing is just documenting it properly and avoid some incorrect syncs where we can?

Reply:
>  it seems that the assertion being made is that this taint goes on nodes which are ALREADY EMPTY, so we can/should treat this the same as a node deletion, I think?

That's not how I interpret the code. I see two cases:

a) "for empty nodes: taint the nodes and asynchronously also delete them - since the nodes are already empty: there's nothing on any of these nodes that might move to another one while we delete them"

b) "for non-empty nodes: taint all the nodes and _only after that is done_ drain them and asynchronously delete them - since the nodes are not empty: we want to avoid having pods move from one node to another node in this same set" 

So, for a) we are fine with just deleting the node from the LB set directly - since they are empty there's nothing to connection drain in any case. For b) we will however abruptly kill all connections for the pods that might still be running on these nodes when we receive the taint update in the CCM.

Reply:
To summarize the options:

CCM options:

1) Keep this taint in the CCM as-is, possibly abruptly killing all connections on a non-drained node but not having the CCM experience any transient errors. 
2) Remove this taint from the CCM, which will cause it to experience transient errors for some time due to syncing LBs with non-existent instances.

service proxy options:

1) we could connection drain a node that is being drained of endpoints simultaneously. But that would involve moving the predicate to the service proxy and failing the HC from the moment we see this taint (if the service proxy is even running at that moment _and_ we receive an HC probe between receiving the node update and actually getting killed ourselves....)
2) do nothing